did/key/key.go

```diff
@@ -90,176 +124,265 @@ func CreateDIDKey(kt crypto.KeyType, publicKey []byte) (*DIDKey, error) {
+	if (publicKeyFormat == cryptosuite.JSONWebKey2020Type && verificationMethod.PublicKeyJWK.CRV == string(cryptosuite.X25519)) ||
+		(publicKeyFormat == cryptosuite.MultikeyType && (verificationMethod.Type == cryptosuite.X25519KeyAgreementKey2020 ||
+			verificationMethod.Type == cryptosuite.X25519KeyAgreementKey2019)) {
+		x25519Key = true
+		doc.Authentication = nil
+		doc.AssertionMethod = nil
+		doc.CapabilityDelegation = nil
+		doc.CapabilityInvocation = nil
+		doc.KeyAgreement = []did.VerificationMethodSet{keyID}
 	}
 
-	if encoding != did.Base58BTCMultiBase {
-		return nil, "", "", fmt.Errorf("expected %d encoding but found %d", did.Base58BTCMultiBase, encoding)
+	// https://w3c-ccg.github.io/did-method-key/#derive-encryption-key-algorithm
+	// the only case we have to consider is if the verification method is X25519
+	if enableEncryptionDerivation && !x25519Key {
+		keyAgreementVerificationMethod, keyAgreementVerificationMethodSet, err := generateKeyAgreementVerificationMethod(*verificationMethod)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not generate key agreement verification method")
+		}
+		if keyAgreementVerificationMethod != nil {
+			doc.VerificationMethod = append(doc.VerificationMethod, *keyAgreementVerificationMethod)
+		}
+		doc.KeyAgreement = keyAgreementVerificationMethodSet
 	}
 
-	// n = # bytes for the int, which we expect to be two from our multicodec
-	multiCodec, n, err := varint.FromUvarint(decoded)
-	if err != nil {
-		return nil, "", "", err
-	}
-	if n != 2 {
-		return nil, "", "", errors.New("error parsing did:key varint")
-	}
+	return &doc, nil
+}
 
-	pubKeyBytes := decoded[n:]
-	multiCodecValue := multicodec.Code(multiCodec)
-	ldKeyType, err := did.MultiCodecToLDKeyType(multiCodecValue)
-	if err != nil {
-		return nil, "", "", errors.Wrap(err, "codec to ld key type")
+func generateKeyAgreementVerificationMethod(vm did.VerificationMethod) (*did.VerificationMethod, []did.VerificationMethodSet, error) {
+	var verificationMethod *did.VerificationMethod
+	var verificationMethodSet []did.VerificationMethodSet
+	var vmErr error
+	if vm.Type == cryptosuite.Ed25519VerificationKey2018 || vm.Type == cryptosuite.Ed25519VerificationKey2020 {
+		// convert key to X25519
+		base58PubKey, err := base58.Decode(vm.PublicKeyBase58)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "could not decode base58 public key")
+		}
+		ed25519PubKey, err := crypto.BytesToPubKey(base58PubKey, crypto.Ed25519)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "could not convert base58 public key to ed25519")
+		}
+		id, x25519Key, err := x25519KeyAndID(vm.Controller, ed25519PubKey.(ed25519.PublicKey))
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "could not generate x25519 key and id")
+		}
+		verificationMethod, vmErr = did.ConstructMultibaseVerificationMethod(id, vm.Controller, x25519Key, cryptosuite.X25519KeyAgreementKey2020)
+		verificationMethodSet = []did.VerificationMethodSet{id}
```

Comment:
For readability, return after the end of each if block. 
```suggestion
		verificationMethodSet = []did.VerificationMethodSet{id}
		return verificationMethod, verificationMethodSet, vmErr
```